File: src/mcp-server/tools/network.py
import subprocess
import shutil

async def get_info() -> str:
    """Get ip addr output"""
    try:
        return subprocess.check_output(["ip", "-brief", "address"]).decode()
    except Exception as e:
        return f"Error getting network info: {e}"

async def download_file(url: str, dest: str) -> str:
    """Download file using curl or wget"""
    if shutil.which("curl"):
        cmd = ["curl", "-L", "-o", dest, url]
    elif shutil.which("wget"):
        cmd = ["wget", "-O", dest, url]
    else:
        return "Error: neither curl nor wget found."
        
    try:
        # This will block the server! For large files we need async execution or a background job.
        # But for now, let's run it. In production, this should be a background task with progress reporting.
        subprocess.run(cmd, check=True)
        return f"Successfully downloaded {url} to {dest}"
    except subprocess.CalledProcessError as e:
        return f"Error downloading file: {e}"

File: src/lib/mcp/tools/network.ts
/* SPDX-License-Identifier: LGPL-2.1-or-later */
import cockpit from "cockpit";

export async function getInfo(): Promise<string> {
    try {
        return await cockpit.spawn(["ip", "-brief", "address"]);
    } catch (e: unknown) {
        return `Error getting network info: ${e instanceof Error ? e.message : String(e)}`;
    }
}

export async function downloadFile(url: string, dest: string): Promise<string> {
    let cmd: string[];

    try {
        await cockpit.spawn(["which", "curl"]);
        cmd = ["curl", "-L", "-o", dest, url];
    } catch {
        try {
            await cockpit.spawn(["which", "wget"]);
            cmd = ["wget", "-O", dest, url];
        } catch {
            return "Error: neither curl nor wget found.";
        }
    }

    try {
        await cockpit.spawn(cmd);
        return `Successfully downloaded ${url} to ${dest}`;
    } catch (e: unknown) {
        return `Error downloading file: ${e instanceof Error ? e.message : String(e)}`;
    }
}
